### ISSUE TITLE
`#[fixture]`: Weird behaviour with `where`

### ISSUE THREAD

**Issue by aliemjay** [la10736/rstest#116]
Using `#[fixture]` with `where` syntax yields some errors with unclear error messages.
Here is my test code: 

``` rust
// rustc 1.51.0
// rstest 0.7.0

use rstest::{fixture, rstest};  

#[fixture]
fn args() -> Vec<i32> {Default::default()}

#[fixture]
fn func<T, S>(args: T)
where
    T: IntoIterator<Item = S>, /* error[E0412]: cannot find type 'S' in this scope */
    S: Copy,
{
}

#[fixture]
fn func2<T>(args: T)
where
    T: IntoIterator,
    T::Item: Copy,  /* error[E0433]: failed to resolve: use of undeclared type 'T' */
{
}

/* However, an equivalent single-line 'where' specs works fine */
#[fixture]
fn func3<S>(args: impl IntoIterator<Item = S>)
where
    S: Copy,
{
}

/* #[rstest] works fine */
#[rstest]
fn func<T, S>(args: T)
where
    T: IntoIterator<Item = S>,
    S: Copy,
{
}
```

**Comment by la10736** [comment]
Ok, thanks to report it: you catch a wired case.  

I need to clean where clause to remove unused types for partial arguments implementations. I should take in account when a type is mentioned just in type constrains and not explicit in function arguments. In this case `S` is implicit in T and cannot be removed.

Ok I'll add this case to may test and try to fix it.

**Comment by aliemjay** [comment]
Great! and thanks again for this great crate!

**Comment by la10736** [comment]
Unfortunately I cannot fix it for the next release (I'm releasing right now) because the fix need some develop effort and I should just release an hot fix and a new feature ... I hope to fix it ASAP in the next weeks.

**Comment by la10736** [comment]
Ok, there are two issues here in the code that clean the signature (we should do it to create the partials impls). We perform clean action also in the base impl and that exposed the issue in some simpler cases.

1. `S` will be discarded because we cannot recognize the transitive dependencies: we don't use `S` directly in the argument list or in the return type but `S` is part of `T` that is used as input argument type.
2. when we implement `func2()` default impl the code don't remove the `T::Item: Copy` where clause because `T != T::Item`

**Comment by la10736** [comment]
Ok I'va also noted that we miss follow:

- [ ] const generics
- [ ] lifetime attribute

I've no idea if a lifetime attribute can be useful for fixture... but I'll implement it

**Comment by la10736** [comment]
Fixed in 0.10.0 Released today

**Comment by aliemjay** [comment]
Thanks you for your time! Very much appreciate it. 

I've tried to understand the code, now my brain hurts :)
